alc/alu.cpp

```diff
@@ -2205,3 +2205,29 @@ void DeviceBase::handleDisconnect(const char *msg, ...)
     }
     IncrementRef(MixCount);
 }
+
+void DeviceBase::handleDefaultChanged(std::string_view deviceId, uint dataFlow)
+{
+    if (Connected.load(std::memory_order_acquire))
+    {
+        AsyncEvent evt{std::in_place_type<AsyncDefaultDeviceChangedEvent>};
+        auto& default_changed = std::get<AsyncDefaultDeviceChangedEvent>(evt);
+        auto len              = std::min(deviceId.length(), sizeof(default_changed.mDeviceId) - 1);
+        if (len > 0)
+            memcpy(default_changed.mDeviceId, deviceId.data(), len);
+        default_changed.mDeviceId[len] = '\0';
+        default_changed.mDataFlow      = dataFlow;
+
+        for (ContextBase* ctx : *mContexts.load())
+        {
+            RingBuffer* ring{ctx->mAsyncEvents.get()};
+            auto evt_data = ring->getWriteVector().first;
```

Comment:
Unfortunately it's not that simple. The ringbuffer is single-producer/single-consumer and is primarily used by the mixer to signal source events as they happen, so you can't write events to the ringbuffer outside of the mixer thread while it's running. I've been trying to find a lockless multi-producer/single-consumer queue, but the things I keep running into sneak in a spinlock on the producer side, making it not lockless and not suitable for the mixer to use.

Reply:
It's need your help, you can write this pr branch of my fork directly.

Reply:
It may be best to not bother with default device change detection for now, since I'm not sure how to do it yet. If I can't get a real-time-safe multi-producer queue working, I may need to do some kind of event marshaling to the mixer which can then put it in the event queue (which isn't straightforward, since the mixer may not be running).

Alternatively, I may set up a separate ALC event API for system-level events, which doesn't have to contend with AL source events being generated in real-time (and could also better handle race conditions, e.g. the app opening the default device, the default device changes, then the app sets up the callback to get event notifications, leaving the app unaware that the default device changed).

Reply:
Maybe add a flag(indicate whether default device changed) to backend and check it in  mixer thread and fire the event?

Reply:
That's one possibility. Though it would need to handle the case of the mixer not running (either because there's no context, the device has been paused, or is disconnected), which can start or stop asynchronously to receiving the system event.

Another possibility is to use the debug API, which isn't as restricted with adding entries to it. It's not as much intended to indicate specific system events, but it's not out of the question.

But I'm thinking a separate ALC API would be better for handling system-generated events, leaving AL_SOFT_events for events generated from mixing. Being in the ALC API would also allow tracking the events without an active device or context, avoiding a race condition.

Reply:
> But I'm thinking a separate ALC API would be better for handling system-generated events, leaving AL_SOFT_events for events generated from mixing. Being in the ALC API would also allow tracking the events without an active device or context, avoiding a race condition.

OK, I have done my changes on branch [wasapi-uwp-support](https://github.com/halx99/openal-soft/tree/wasapi-uwp-support)

Reply:
I have done notify default device change by `alc::Event`